plugins/fabric8-multi-tenant-manager/src/main/java/com/redhat/che/multitenant/pvc/CommonPVCStrategyWithEphemeralWorkspacesSupport.java

```diff
@@ -0,0 +1,167 @@
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalInfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.Names;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.PVCSubPathHelper;
+
+/**
+ * Extends {@code CommonPVCStrategy} and provides support of the ephemeral workspaces based on the
+ * `emptyDir` volumes. If workspace config contains `mountSources` attribute which is set to true,
+ * than instead of using regular {@code CommonPVCStrategy}, each volume would be represented as an
+ * `emptyDir`. NOTE: no data would be persisted once between ephemeral workspace restarts, since
+ *
+ * @see <a href="https://kubernetes.io/docs/concepts/storage/volumes/#emptydir">EmptyDir</a>
+ * @author Ilya Buziuk
+ */
+public class CommonPVCStrategyWithEphemeralWorkspacesSupport extends CommonPVCStrategy {
+  public static final String COMMON_EPHEMERAL_STRATEGY = "common_ephemeral";
+  private static final String MOUNT_SOURCES_ATTRIBUTE = "mountSources";
+
+  private final WorkspaceManager workspaceManager;
+  private final String volumeNamePrefix;
+
+  @Inject
+  public CommonPVCStrategyWithEphemeralWorkspacesSupport(
+      @Named("che.infra.kubernetes.pvc.name") String pvcName,
+      @Named("che.infra.kubernetes.pvc.quantity") String pvcQuantity,
+      @Named("che.infra.kubernetes.pvc.access_mode") String pvcAccessMode,
+      @Named("che.infra.kubernetes.pvc.precreate_subpaths") boolean preCreateDirs,
+      PVCSubPathHelper pvcSubPathHelper,
+      KubernetesNamespaceFactory factory,
+      WorkspaceManager workspaceManager) {
+    super(pvcName, pvcQuantity, pvcAccessMode, preCreateDirs, pvcSubPathHelper, factory);
+    this.workspaceManager = workspaceManager;
+    this.volumeNamePrefix = pvcName;
+  }
+
+  @Override
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    String workspaceId = identity.getWorkspaceId();
+    if (isEphemeral(workspaceId)) {
+      for (Pod pod : k8sEnv.getPods().values()) {
+        final PodSpec podSpec = pod.getSpec();
+        for (Container container : podSpec.getContainers()) {
+          final String machineName = Names.machineName(pod, container);
+          Map<String, Volume> volumes = k8sEnv.getMachines().get(machineName).getVolumes();
+          addMachineVolumes(workspaceId, pod, container, volumes);
+        }
+      }
+    } else {
+      super.provision(k8sEnv, identity);
+    }
+  }
+
+  @Override
+  public void cleanup(String workspaceId) throws InfrastructureException {
+    super.cleanup(workspaceId);
```

Comment:
@garagatyi @sleshchenko guys, it is appeared to be that in the strategy `cleanup` method it is not possible to check if the `mountSources` attribute is presented, since workspace is already deleted. Do you have any ideas how ephemeral / non-ephemeral check should be performed in this case, or for the time being it would be ok if `cleanup` would still be called once ephemeral workspace is deleted (the operation is async and in case of ephemeral workspace the `rm-*` pod would just not do any real cleanup operations against pvc)

Reply:
Here is the code that calls cleanup [link](https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/WorkspacePVCCleaner.java#L59)
Here where it takes workspace ID from [link](https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/WorkspacePVCCleaner.java#L57)
Here you can see that this object contains whole workspace object and this means that we can pass whole workspace instead of just an ID [link](https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/wsmaster/che-core-api-workspace-shared/src/main/java/org/eclipse/che/api/workspace/shared/event/WorkspaceRemovedEvent.java#L26)

Hope it helps.

Reply:
thanks, but I was not able to override `WorkspacePVCCleaner` in our assembly so far :-(